packages/microcosm/src/domain-engine.js

```diff
@@ -112,60 +127,78 @@ class DomainEngine {
 
   dispatch(action: Action, state: Object, snapshot: Snapshot) {
     let handlers = this.register(action)
-    let result = state
+    let answer = state
 
     for (var i = 0; i < handlers.length; i++) {
-      var { key, scope, steps } = handlers[i]
+      var { local, key, scope, steps } = handlers[i]
 
-      var base = get(result, key, null)
-      var head = get(snapshot.last, key, null)
+      if (local && action.origin !== this.repo) {
+        continue
+      }
+
+      var last = answer[key]
+      var head = snapshot.last[key]
+      var next = snapshot.next[key]
 
       if (
         // If the reference to the prior state changed
-        base !== head ||
+        last !== head ||
         // Or the payload is different
         action.payload !== snapshot.payload ||
         // or the status is different
         action.status !== snapshot.status
       ) {
-        // Yes: recalculate state from the base
-        result = set(result, key, reduce(steps, action.payload, base, scope))
-      } else {
-        // No: use the existing snapshot value (memoizing the domain handler)
-        result = set(result, key, get(snapshot.next, key))
+        // Recalculate state from the last answer
+        next = reduce(steps, action.payload, last, scope)
+      }
+
+      if (answer === state && next !== answer[key]) {
+        answer = clone(state)
       }
```

Comment:
can you summarize what this conditional is checking, and then doing if it returns true?

Reply:
Yep! This check is saying "If the answer object reference is the same as the state object reference, and the value has changed, then clone the answer".

This is pretty low level. Basically I want to avoid shallow cloning the next state object every time a domain handler runs.